// @polymedia/suitcase-react/dist/cjs/networks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkRadioSelector = NetworkRadioSelector;
exports.NetworkDropdownSelector = NetworkDropdownSelector;
exports.isLocalhost = isLocalhost;
exports.loadNetwork = loadNetwork;
exports.switchNetwork = switchNetwork;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const hooks_1 = require("./hooks");
const selectors_1 = require("./selectors");
/**
 * A radio button menu to select a Sui network and save the choice to local storage.
 */
function NetworkRadioSelector(props) {
    const options = props.supportedNetworks.map(network => ({
        value: network,
        label: network
    }));
    const onSelect = (newNetwork) => {
        switchNetwork(newNetwork, props.supportedNetworks, props.onSwitch);
    };
    return ((0, jsx_runtime_1.jsx)(selectors_1.RadioSelector, { options: options, selectedValue: props.selectedNetwork, onSelect: onSelect, className: `poly-network-radio-selector ${props.className ?? ""}` }));
}
/**
 * A dropdown menu to choose between mainnet/testnet/devnet/localnet.
 */
function NetworkDropdownSelector(props) {
    const [isOpen, setIsOpen] = (0, react_1.useState)(false);
    const selectorRef = (0, react_1.useRef)(null);
    (0, hooks_1.useClickOutside)(selectorRef, () => { setIsOpen(false); });
    const SelectedOption = () => {
        return (0, jsx_runtime_1.jsx)("div", { className: "network-option selected" /* onMouseEnter={() => setIsOpen(true)} */, children: (0, jsx_runtime_1.jsx)("span", { className: "text", onClick: () => { !props.disabled && setIsOpen(true); }, children: props.currentNetwork }) });
    };
    const NetworkOptions = () => {
        const otherNetworks = props.supportedNetworks.filter(net => net !== props.currentNetwork);
        return (0, jsx_runtime_1.jsx)("div", { className: "network-options", children: otherNetworks.map(net => ((0, jsx_runtime_1.jsx)(NetworkOption, { network: net }, net))) });
    };
    const NetworkOption = ({ network }) => {
        return (0, jsx_runtime_1.jsx)("div", { className: "network-option", children: (0, jsx_runtime_1.jsx)("span", { className: "text", onClick: () => {
                    if (!props.disabled) {
                        switchNetwork(network, props.supportedNetworks, props.onSwitch);
                        setIsOpen(false);
                    }
                }, children: network }) });
    };
    return (0, jsx_runtime_1.jsxs)("div", { id: props.id, className: "network-selector " + (props.disabled ? "disabled " : "") + (props.className ?? ""), ref: selectorRef, onMouseLeave: () => { setIsOpen(false); }, children: [(0, jsx_runtime_1.jsx)(SelectedOption, {}), isOpen && (0, jsx_runtime_1.jsx)(NetworkOptions, {})] });
}
/**
 * Check if the current hostname is a localhost environment.
 */
function isLocalhost() {
    const hostname = window.location.hostname;
    const localNetworkPattern = /(^127\.)|(^10\.)|(^172\.1[6-9]\.)|(^172\.2[0-9]\.)|(^172\.3[0-1]\.)|(^192\.168\.)/;
    return hostname === "localhost" || localNetworkPattern.test(hostname);
}
/**
 * Load the network name based on URL parameters and local storage.
 */
function loadNetwork(supportedNetworks, defaultNetwork) {
    if (!isNetworkName(defaultNetwork, supportedNetworks)) {
        throw new Error(`Network not supported: ${defaultNetwork}`);
    }
    // Use 'network' URL parameter, if valid
    const params = new URLSearchParams(window.location.search);
    const networkFromUrl = params.get("network");
    if (isNetworkName(networkFromUrl, supportedNetworks)) {
        params.delete("network");
        const newQuery = params.toString();
        const newUrl = window.location.pathname + (newQuery ? "?" + newQuery : "") + window.location.hash;
        window.history.replaceState({}, document.title, newUrl);
        localStorage.setItem("polymedia.network", networkFromUrl);
        return networkFromUrl;
    }
    // Use network from local storage, if valid
    const networkFromLocal = localStorage.getItem("polymedia.network");
    if (isNetworkName(networkFromLocal, supportedNetworks)) {
        return networkFromLocal;
    }
    // Use default network
    localStorage.setItem("polymedia.network", defaultNetwork);
    return defaultNetwork;
}
/**
 * Change networks, update local storage, and optionally trigger a callback.
 */
function switchNetwork(newNetwork, supportedNetworks, onSwitch) {
    if (!isNetworkName(newNetwork, supportedNetworks)) {
        throw new Error(`Network not supported: ${newNetwork}`);
    }
    localStorage.setItem("polymedia.network", newNetwork);
    if (onSwitch) {
        onSwitch(newNetwork);
    }
    else {
        window.location.reload();
    }
}
function isNetworkName(value, supportedNetworks) {
    return value !== null && supportedNetworks.includes(value);
}
