// @polymedia/suitcase-core@0.0.57/dist/cjs/guards.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isOwnerKind = isOwnerKind;
exports.isArgKind = isArgKind;
exports.isObjChangeKind = isObjChangeKind;
exports.isSuiObjectRef = isSuiObjectRef;
exports.isParsedDataKind = isParsedDataKind;
exports.isTxKind = isTxKind;
/**
 * Type guard to check if an `ObjectOwner` is of a specific kind.
 */
function isOwnerKind(owner, kind) {
    return kind === owner || (typeof owner === "object" && kind in owner);
}
/**
 * Type guard to check if a `SuiArgument` is of a specific kind.
 */
function isArgKind(arg, kind) {
    return kind === arg || (typeof arg === "object" && kind in arg);
}
/**
 * Type guard to check if a `SuiObjectChange` is of a specific kind.
 */
function isObjChangeKind(change, kind) {
    return change.type === kind;
}
// === SuiObjectRef ===
/** Type guard to check if an object is a `SuiObjectRef`. */
function isSuiObjectRef(obj) {
    return typeof obj === "object" && obj !== null
        && "objectId" in obj
        && "version" in obj
        && "digest" in obj;
}
/**
 * Type guard to check if a `SuiParsedData` is of a specific kind.
 */
function isParsedDataKind(data, kind) {
    return data.dataType === kind;
}
/**
 * Type guard to check if a `SuiTransaction` is of a specific kind.
 */
function isTxKind(tx, kind) {
    return kind in tx;
}
