// @polymedia/.ignored_suitcase-core/dist/esm/guards.js
/**
 * Type guard to check if an `ObjectOwner` is of a specific kind.
 */
export function isOwnerKind(owner, kind) {
    return kind === owner || (typeof owner === "object" && kind in owner);
}
/**
 * Type guard to check if a `SuiArgument` is of a specific kind.
 */
export function isArgKind(arg, kind) {
    return kind === arg || (typeof arg === "object" && kind in arg);
}
/**
 * Type guard to check if a `SuiObjectChange` is of a specific kind.
 */
export function isObjChangeKind(change, kind) {
    return change.type === kind;
}
// === SuiObjectRef ===
/** Type guard to check if an object is a `SuiObjectRef`. */
export function isSuiObjectRef(obj) {
    return typeof obj === "object" && obj !== null
        && "objectId" in obj
        && "version" in obj
        && "digest" in obj;
}
/**
 * Type guard to check if a `SuiParsedData` is of a specific kind.
 */
export function isParsedDataKind(data, kind) {
    return data.dataType === kind;
}
/**
 * Type guard to check if a `SuiTransaction` is of a specific kind.
 */
export function isTxKind(tx, kind) {
    return kind in tx;
}
